ultravox/tools/infer_tool.py

```diff
@@ -190,25 +192,48 @@ def dataset_infer(inference: base.VoiceInference, args: InferArgs):
     if args.seed is not None:
         ds_args.shuffle_seed = args.seed
     ds = datasets.create_dataset(args.data_sets[0], ds_args)
-    scores: List[float] = []
-    for i, sample in enumerate(datasets.Range(ds, args.num_samples)):
-        # Store the original question and answer for JSON output.
-        question_text = sample.audio_transcript
-        expected_answer = sample.messages[-1]["content"]
-        # Drop any assistant response from the sample.
-        sample.messages = sample.messages[:-1]
-        if not args.json:
-            run_tui(i, inference, sample, args, expected_answer, scores)
-        else:
-            output = inference.infer(
-                sample, max_tokens=args.max_tokens, temperature=args.temperature
+
+    if args.json and isinstance(inference, infer.LocalInference):
+        start_time = time.time()
+        dl = DataLoader(
+            datasets.Range(ds, args.num_samples),
+            batch_size=args.batch_size,
+            collate_fn=lambda x: x,
+        )
+        current_batch = []
+        sample_index = 0
+        for current_batch in dl:
+            output = []
+            for sample in current_batch:
+                output.append(
+                    {
+                        "index": sample_index,
+                        "question_text": sample.audio_transcript,
+                        "expected_answer": sample.messages[-1]["content"],
+                    }
+                )
+                sample.messages = sample.messages[:-1]
```

Comment:
How do we handle the cases where the dataset doesn't have reference for the samples?

Reply:
Ah the code is sort of hardcoded to expect that right now. Do think we should make that change _after_ we start seeing datasets that don't have reference in the samples? Else I'd imagine we could add a flag for that and have another line of logic to handle the other case.

Reply:
Sounds good. I can imagine two use cases.

1. generating synthesized data for model training
2. self-evaluation: comparing SLM's output against LLM's output, where no reference is available.

Let's try to merge this PR first.